fs.go

```diff
@@ -208,3 +210,104 @@ func CopyFile(src string, dest string) error {
 
 	return os.Chmod(dest, srcinfo.Mode())
 }
+
+// hasFilepathPrefix will determine if "path" starts with "prefix" from
+// the point of view of a filesystem.
+//
+// Unlike strings.HasPrefix, this function is path-aware, meaning that
+// it knows that two directories /foo and /foobar are not the same
+// thing, and therefore hasFilepathPrefix("/foobar", "/foo") will return
+// false.
+func hasFilepathPrefix(path, prefix string) bool {
+	if filepath.VolumeName(path) != filepath.VolumeName(prefix) {
+		return false
+	}
+
+	var dn string
+
+	if isDir, err := IsDir(path); err != nil {
+		return false
+	} else if isDir {
+		dn = path
+	} else {
+		dn = filepath.Dir(path)
+	}
+
+	dn = strings.TrimSuffix(dn, string(os.PathSeparator))
+	prefix = strings.TrimSuffix(prefix, string(os.PathSeparator))
+
+	dirs := strings.Split(dn, string(os.PathSeparator))[1:]
+	prefixes := strings.Split(prefix, string(os.PathSeparator))[1:]
+
+	if len(prefixes) > len(dirs) {
+		return false
+	}
+
+	var d, p string
+
+	for i := range prefixes {
+		// need to test each component of the path for
+		// case-sensitiveness because on Unix we could have
+		// something like ext4 filesystem mounted on FAT
+		// mountpoint, mounted on ext4 filesystem, i.e. the
+		// problematic filesystem is not the last one.
+		if isCaseSensitiveFilesystem(filepath.Join(d, dirs[i])) {
+			d = filepath.Join(d, dirs[i])
+			p = filepath.Join(p, prefixes[i])
+		} else {
+			d = filepath.Join(d, strings.ToLower(dirs[i]))
+			p = filepath.Join(p, strings.ToLower(prefixes[i]))
+		}
+
+		if p != d {
+			return false
+		}
+	}
+
+	return true
+}
+
+// flipCase returns an string with all the runes case-flipped
+func flipCase(str string) string {
+	return strings.Map(func(r rune) rune {
```

Comment:
Soooo...I'm reading more about unicode rules, and apparently, [it's also not a guarantee that upper or lower-casing an individual character is a reversible operation](http://unicode.org/faq/casemap_charprop.html#7a).

It gets even less fun, too: https://play.golang.org/p/u1TO6yMn7F

How does this sound:

1. Instead of trying to case-flip ALL the characters in the string, just proceed until we find one that's flippable, flip that one, and be done.
2. If `unicode.IsUpper()` says yes, then see if the flip is reversible, `unicode.ToUpper(unicode.ToLower(r))`; if we don't have the same original input, then skip to the next rune. So:

```go
if unicode.IsUpper(r) {
	l := unicode.ToLower(r)
	if r == unicode.ToUpper(l) {
		// we've flipped case reversibly, so signal to stop somehow here
		return l
	}
	return r
}
```

2. Do the same pattern of check for lower case
3. If we weren't able to flip case at all, return false (as you do now, for the same reasons you do now)